src/Confluent.Kafka.Avro/AvroConsumer.cs

```diff
@@ -0,0 +1,201 @@
+        /// </summary>
+        /// <param name="schemaRegistryClient">
+        ///     An implementation of ISchemaRegistryClient used for
+        ///     communication with Confluent Schema Registry.
+        /// </param>
+        /// <param name="config">
+        ///     A collection of librdkafka configuration properties.
+        /// </param>
+        public AvroConsumer(ISchemaRegistryClient schemaRegistryClient, IEnumerable<KeyValuePair<string, string>> config)
+            : base(config)
+        {
+            this.schemaRegistryClient = schemaRegistryClient;
+        }
+
+
+        /// <summary>
+        ///     Sets the Avro deserializer that will be used to deserialize keys or values with
+        ///     the specified type.
+        /// </summary>
+        /// <param name="deserializer">
+        ///     The avro deserializer.
+        /// </param>
+        public void RegisterAvroDeserializer<T>(AvroDeserializer<T> deserializer)
+        {
+            avroDeserializers.Add(typeof(T), deserializer);
+        }
+
+
+        /// <summary>
+        ///     Removes the avro deserializer associated with the specified type.
+        /// </summary>
+        public void UnregisterAvroSerializer<T>()
+        {
+            avroDeserializers.Remove(typeof(T));
+        }
+
+
+        /// <summary>
+        ///     Gets the avro deserializer that will be used to deserialize values of the specified type.
+        /// </summary>
+        /// <returns>
+        ///     The avro deserializer corresponding to the specified type.
+        /// </returns>
+        public AvroDeserializer<T> GetAvroDeserializer<T>()
+        {
+            return (AvroDeserializer<T>)avroDeserializers[typeof(T)];
+        }
+
+
+        private AvroDeserializer<T> GetOrCreateAvroDeserializer<T>()
+        {
+            if (!avroDeserializers.ContainsKey(typeof(T)))
+            {
+                avroDeserializers.Add(typeof(T), new AvroDeserializer<T>());
+            }
+
+            return (AvroDeserializer<T>)avroDeserializers[typeof(T)];
+        }
+
+
+        /// <summary>
+        ///     Poll for new messages / events. Blocks until a 
+        ///     <see cref="Confluent.Kafka.ConsumeResult{TKey, TValue}" />
+        ///     is available or the operation has been cancelled.
+        /// </summary>
+        /// <param name="keySerdeType">
+        ///     The deserializer type to use to deserialize keys.
+        /// </param>
+        /// <param name="valueSerdeType">
+        ///     The deserializer type to use to deserialize values.
+        /// </param>
+        /// <param name="cancellationToken">
+        ///     A cancellation token that can be used to cancel this operation.
+        /// </param>
+        /// <returns>
+        ///     The <see cref="Confluent.Kafka.ConsumeResult{TKey, TValue}" />.
+        /// </returns>
+        public async Task<ConsumeResult<TKey, TValue>> ConsumeAsync<TKey, TValue>(
```

Comment:
I didn't investigate this exactly, but the performance of `ProduceAsync` is considerably worse than `BeginProduce` (IIRC something like 30%) and the main difference is creation of `Task` objects, so it's significant. Here we also have the thread pool coming into play via `Task.Run`, which is probably worse.

HOWEVER, you've pointed out something I think I forgot to change. On the producer side, I previously had `BeginProduce` async (so as to not block on the SR call), but this doesn't make sense given how people will use it, so I changed it to block on a background thread. The same applies here I think but the analogous change is to just block.

This is a little bit fraught with danger (see reported deadlock bugs related with avro producer in winforms apps), but I think it's ok since no one will be ever making a blocking consume call where it would matter.